cpp/src/arrow/ipc/message.h

```diff
@@ -163,6 +164,195 @@ class ARROW_EXPORT Message {
+  ///
+  /// \param[in] data a raw data to be processed. This data isn't
+  /// copied. The passed memory must be kept alive through message
+  /// processing.
+  /// \param[in] size raw data size.
+  /// \return Status
+  Status Consume(const uint8_t* data, int64_t size);
+
+  /// \brief Feed data to the emitter as a Buffer.
+  ///
+  /// If the emitter can read one or more messages by the Buffer, the
+  /// emitter emits read messages by calling receiver->Receive()
+  /// multiple times.
+  ///
+  /// \param[in] buffer a Buffer to be processed.
+  /// \return Status
+  Status Consume(std::shared_ptr<Buffer> buffer);
+
+  /// \brief Return the number of bytes needed to advance the state of
+  /// the emitter.
+  ///
+  /// This method is provided for users who want to optimize performance.
+  /// Normal users don't need to use this method.
+  ///
+  /// Here is an example usage for normal users:
+  ///
+  /// ~~~{.cpp}
+  /// emitter.Consume(buffer1);
+  /// emitter.Consume(buffer2);
+  /// emitter.Consume(buffer3);
+  /// ~~~
+  ///
+  /// Emitter has internal buffer. If consumed data isn't enough to
+  /// advance the state of the emitter, consumed data is buffered to
+  /// the internal buffer. It causes performance overhead.
+  ///
+  /// If you pass next_required_size() size data to each Consume()
+  /// call, the emitter doesn't use its internal buffer. It improves
+  /// performance.
+  ///
+  /// Here is an example usage to avoid using internal buffer:
+  ///
+  /// ~~~{.cpp}
+  /// buffer1 = get_data(emitter.next_required_size());
+  /// emitter.Consume(buffer1);
+  /// buffer2 = get_data(emitter.next_required_size());
+  /// emitter.Consume(buffer2);
+  /// ~~~
+  ///
+  /// Users can use this method to avoid creating small
+  /// chunks. Message body must be contiguous data. If users pass
+  /// small chunks to the emitter, the emitter needs concatenate small
+  /// chunks internally. It causes performance overhead.
+  ///
+  /// Here is an example usage to reduce small chunks:
+  ///
+  /// ~~~{.cpp}
+  /// buffer = AllocateResizableBuffer();
+  /// while ((small_chunk = get_data(&small_chunk_size))) {
+  ///   auto current_buffer_size = buffer->size();
+  ///   buffer->Resize(current_buffer_size + small_chunk_size);
+  ///   memcpy(buffer->mutable_data() + current_buffer_size,
+  ///          small_chunk,
+  ///          small_chunk_size);
+  ///   if (buffer->size() < emitter.next_requied_size()) {
+  ///     continue;
+  ///   }
+  ///   std::shared_ptr<arrow::Buffer> chunk(buffer.release());
+  ///   emitter.Consume(chunk);
+  ///   buffer = AllocateResizableBuffer();
+  /// }
+  /// if (buffer->size() > 0) {
+  ///   std::shared_ptr<arrow::Buffer> chunk(buffer.release());
+  ///   emitter.Consume(chunk);
+  /// }
+  /// ~~~
+  ///
+  /// \return the number of bytes needed to advance the state of the
+  /// emitter
+  int64_t next_required_size() const;
+
+  /// \return the current state
+  State state() const;
```

Comment:
Instead of this, I think it would be better to have an `EosReceived` method on `MessageReceiver`.

Reply:
I've added EOS callback but I want to keep this method.
Because this information can be used to optimize performance. I've added documentation for this.